django/transformer/reference_binder.py

```diff
@@ -0,0 +1,97 @@
+import logging
+from typing import List
+from uuid import UUID, uuid5
+
+from dotty_dict import dotty
+from loader.load.utils import get_resource_id
+
+logger = logging.getLogger(__name__)
+
+
+class ReferenceBinder:
+    def resolve_references(self, unresolved_fhir_object, reference_paths: List[List[str]]):
+        fhir_object = dotty(unresolved_fhir_object)
+        resource_id = get_resource_id(unresolved_fhir_object)
+
+        # iterate over the instance's references and try to resolve them
+        for reference_path in reference_paths:
+            logger.debug(
+                {
+                    "message": f"Trying to resolve reference for resource {fhir_object['id']} "
+                    f"at {'[*]'.join(reference_path)}",
+                    "resource_id": resource_id,
+                },
+            )
+            try:
+                self.bind_existing_reference(fhir_object, reference_path)
+            except KeyError as e:
+                logger.warning(f"{reference_path} does not exist in resource {fhir_object['id']}: {e}")
+
+        return fhir_object.to_dict()
+
+    def bind_existing_reference(self, fhir_object, reference_path: List[str]):
+        resource_id = get_resource_id(fhir_object)
+        object_id = fhir_object["id"]
+
+        def bind(ref):
+            # extract the type and itentifier of the reference
+            if not (reference_type := ref.get("type")):
+                logger.debug(f"reference at path {ref} has no type")
+                return
+            if not (identifier := ref.get("identifier")):
+                logger.debug(f"reference at path {ref} has no identifier")
+                return
+
+            # add the "literal reference"
+            # (https://www.hl7.org/fhir/references-definitions.html#Reference.reference)
+            try:
+                ref["reference"] = self.identifier_to_reference(identifier, reference_type)
+                logger.debug(
+                    {
+                        "message": f"reference {ref['reference']} of resource {object_id} resolved",
+                        "resource_id": resource_id,
+                    },
+                )
+            except (ValueError, KeyError) as e:
```

Comment:
Only a `KeyError` can occur if an identifier is incomplete, right?

Reply:
A ValueError can occur if if the value in the system attribute does not contain a valid uuid